src/Fl_Native_File_Chooser_Zenity.cxx

```diff
@@ -0,0 +1,284 @@
+#include "Fl_Native_File_Chooser_Zenity.H"
+#include "Fl_Window_Driver.H"
+#include "drivers/Unix/Fl_Unix_System_Driver.H"
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+
+/* Fl_Zenity_Native_File_Chooser_Driver : file chooser based on the "zenity" command */
+
+bool Fl_Zenity_Native_File_Chooser_Driver::did_find_zenity = false;
+bool Fl_Zenity_Native_File_Chooser_Driver::have_looked_for_zenity = false;
+
+
+Fl_Zenity_Native_File_Chooser_Driver::Fl_Zenity_Native_File_Chooser_Driver(int val) :  Fl_Native_File_Chooser_FLTK_Driver(val) {
+  _tpathnames = 0;
+  _pathnames = NULL;
+  _directory = NULL;
+  _preset_file = NULL;
+  _title = NULL;
+}
+
+
+Fl_Zenity_Native_File_Chooser_Driver::~Fl_Zenity_Native_File_Chooser_Driver() {
+  for (int i = 0; i < _tpathnames; i++) delete[] _pathnames[i];
+  delete[] _pathnames;
+  if (_preset_file) free(_preset_file);
+  if (_directory) free(_directory);
+  if (_title) free(_title);
+}
+
+
+void Fl_Zenity_Native_File_Chooser_Driver::fnfc_fd_cb(int fd,
+                      Fl_Zenity_Native_File_Chooser_Driver::fnfc_pipe_struct *data) {
+  char tmp[FL_PATH_MAX];
+  int l = read(fd, tmp, sizeof(tmp)-1);
+  if (l > 0) {
+    tmp[l] = 0;
+    data->all_files = Fl_Native_File_Chooser_Driver::strapp(data->all_files, tmp);
+  } else {
+    data->fd = -1;
+  }
+}
+
+
+static int fnfc_dispatch(int /*event*/, Fl_Window* /*win*/) {
+  return 0;
+}
+
+
+int Fl_Zenity_Native_File_Chooser_Driver::show() {
+  const char *option;
+  switch (_btype) {
+    case Fl_Native_File_Chooser::BROWSE_MULTI_DIRECTORY: {
+      // BROWSE_MULTI_DIRECTORY is not supported by zenity, run other chooser instead
+      Fl_Native_File_Chooser fnfc(Fl_Native_File_Chooser::BROWSE_MULTI_DIRECTORY);
+      fnfc.title( title() );
+      fnfc.directory(directory());
+      fnfc.preset_file(preset_file());
+      fnfc.filter(filter());
+      fnfc.options(options());
+      int retval = fnfc.show();
+      for (int i = 0; i < _tpathnames; i++) delete[] _pathnames[i];
+      delete[] _pathnames; _pathnames = NULL;
+      _tpathnames = fnfc.count();
+      if (_tpathnames && retval == 0) {
+        _pathnames = new char*[_tpathnames];
+        for (int i = 0; i < _tpathnames; i++) {
+          _pathnames[i] = new char[strlen(fnfc.filename(i))+1];
+          strcpy(_pathnames[i], fnfc.filename(i));
+        }
+      }
+      return retval;
+    }
+      break;
+    case Fl_Native_File_Chooser::BROWSE_DIRECTORY:
+    case Fl_Native_File_Chooser::BROWSE_SAVE_DIRECTORY:
+      option = "--file-selection --directory";
+      break;
+
+    case Fl_Native_File_Chooser::BROWSE_SAVE_FILE:
+      option = "--file-selection --save";
+      break;
+
+    case Fl_Native_File_Chooser::BROWSE_MULTI_FILE:
+      option = "--file-selection --multiple";
+      break;
+
+    default:
+      option = "--file-selection";
+  }
+  char *preset = (char*)"";
+  if (_preset_file) {
+    preset = new char[strlen(_preset_file) + 15];
```

Comment:
This and the cast above looks like causing a memory leak. Perhaps the initialization should be
```
  char *preset = 0; // or NULL
```
and preset should be delete[]'ed somewhere (later).

Reply:
Will initializing it like that still work for the sprintf at line 127? That's where it ultimately gets used, so I think right after that is a good place to delete it.